bazaar/src/lib.rs

```diff
@@ -89,73 +88,113 @@ decl_module! {
-            ensure!(!ShopOwner::<T>::contains_key(cid, &shop), Error::<T>::ShopAlreadyCreated);
+            Self::deposit_event(RawEvent::BusinessCreated(cid, sender));
 
-            // Add the shop to the registries
-            owned_shops.push(shop.clone());
-            shops.push(shop.clone());
-            // Update blockchain
-            ShopsOwned::<T>::insert(cid, &sender, owned_shops);
-            ShopOwner::<T>::insert(cid, &shop, &sender);
-            ShopRegistry::insert(cid, shops);
-            // Emit an event that the shop was created
-            Self::deposit_event(RawEvent::ShopCreated(cid, sender, shop));
             Ok(())
         }
 
-        /// Allow a user to remove their shop
         #[weight = 10_000]
-        pub fn remove_shop(origin, cid:CommunityIdentifier, shop: ShopIdentifier) -> DispatchResult {
+        pub fn update_business(origin, cid: CommunityIdentifier, url: PalletString) -> DispatchResult {
             // Check that the extrinsic was signed and get the signer
             let sender = ensure_signed(origin)?;
+            // Check that the supplied community is actually registered
+            ensure!(<encointer_communities::Module<T>>::community_identifiers().contains(&cid),
+                Error::<T>::InexistentCommunity);
+
+            ensure!(BusinessRegistry::<T>::contains_key(cid, sender.clone()), Error::<T>::InexistentBusiness);
+
+            BusinessRegistry::<T>::mutate(cid, sender.clone(), |b| b.url = url);
+
+            Self::deposit_event(RawEvent::BusinessUpdated(cid, sender));
+
+            Ok(())
+        }
+
+        #[weight = 10_000]
+        pub fn delete_business(origin, cid: CommunityIdentifier) -> DispatchResult {
+            // Check that the extrinsic was signed and get the signer
+            let sender = ensure_signed(origin)?;
+            // Check that the supplied community is actually registered
+            ensure!(<encointer_communities::Module<T>>::community_identifiers().contains(&cid),
+                Error::<T>::InexistentCommunity);
+
+            ensure!(BusinessRegistry::<T>::contains_key(cid, sender.clone()), Error::<T>::InexistentBusiness);
+
+            BusinessRegistry::<T>::remove(cid, sender.clone());
+            OfferingRegistry::<T>::remove_prefix(BusinessIdentifier{community_identifier: cid, business_account: sender.clone()});
+
+            Self::deposit_event(RawEvent::BusinessDeleted(cid, sender));
+
+            Ok(())
+        }
+
+        #[weight = 10_000]
+        pub fn create_offering(origin, cid: CommunityIdentifier, url: PalletString) -> DispatchResult {
+            // Check that the extrinsic was signed and get the signer
+            let sender = ensure_signed(origin)?;
+            // Check that the supplied community is actually registered
+            ensure!(<encointer_communities::Module<T>>::community_identifiers().contains(&cid),
```

Comment:
This check is actually redundant. The next ensure will fail anyhow if the cid does not exist. The same goes for the other methods.